### DISCUSSION TITLE
Recommendation: Process in multi-band mode or write to multiple regions of output image simultaneously (how)?

### DISCUSSION THREAD

**Discussion by prasadbandodkar** [libvips/libvips#4181] [Q&A] [answered]
Hi,

I have a 10-channel, 100k x 100k pyramidal ome.tiff image. When I load this image, each channel appears on a separate page in vips. I want to perform some image analysis operations on it and save the result as a pyramidal ome.tiff image. Here’s my proposed approach: is this the best way to proceed? I need to process a region from all channels simultaneously. 

### Method:
1. Combine all channels as bands and create a 10-band image.
2. Ask for pixels from a region
3. Perform custom image analysis using non-vips library in a multi-band mode
4. Run over all regions
5. Bandsplit the resulting output image
6. Arrayjoin the split images and save a pyramidal ome.tiff image

Other things to note:
- I'll be doing this in C or C++

Specifically I want to define a VipsOperation, where I would define the output image as (10 x image_height) x image_width. Then, I ask for a  region like {0, 0, 128, 128} from all channels as they are individual input images (not multi-band). So I have 10 of these input regions. Now, I want to write the result to 10 different locations within the output image, at {0, 0}, {image_height+1, 0}, {2*image_height+1, 0}, ... and so on. It's necessary for me to process all images simultaneously.

**Comment by jcupitt** [comment]
Hello @prasadbandodkar,

I would do some benchmarks with a smaller image first (of course) and use the numbers to refine my approach. But maybe start with:

- create a full-size intermediate image (100k across, 1m down) on disc with `black` (you'll need 100gb of disc space)
- then open in write mode
- load the input image as one very tall, thin page
- `crop` to a set of separate pages
- `bandjoin` to a 100 band image
- `crop` 128x128 pixels out and write to memory (`fetch` might possibly be quicker, you'd need to test)
- run the analysis on that area of memory and generate a new 128x128x10 patch
- use `new_from_memory` to wrap that patch as a vips image
- `bandsplit` to an array of images
- use `vips_draw_image()` to draw each image at the right position in the 100gb intermediate
- when you've processed every tile, use `tiffsave` to write the 1tb intermediate out as an OME-TIFF

Instead of drawing into a huge disc image, you could also write a set of 128x128x10 tiles out in vips format, then to make the output, use `arrayjoin` and `bandsplit` to assemble the OME-TIFF image for writing. You'd still need 1tb of disc space, but it would be easier to stop and start processing.

You could also write the intermediate as a set of 10 128x128 mono JPEGs and save some disc space that way.

You could also write your own VipsOperation which pumped tiles through your external processing library. Now you wouldn't need a disc intermediate, but you would need to be a C programmer.

**Comment by jcupitt** [comment, marked as answer]
Making your own VipsOperation would probably be the best solution, but not the easiest.

I made a sample plugin here:

https://github.com/jcupitt/vips-gmic

That pipes vips images through G'MIC

https://gmic.eu/

Which is close to what you want.

**Reply by prasadbandodkar** [reply]
Hi @jcupitt , thank you for your prompt response. Both techniques were insightful and useful. I found the vips-gmic example to be the most useful, therefore, am marking it as the answer.
